docs/proposals/ingress-l4.md

```diff
@@ -0,0 +1,320 @@
+
+    - protocol: tls
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+### L4 TLS Termination
+
+While it would be very useful to be able to express TLS protected L4 Ingress,
+such that TLS can be terminated by the Ingress implementation, not all cloud
+providers support TLS termination, thus more design is required to add L4 TLS
+for Ingress in a manner that would work for all cloud providers.
+
+## Proposed Addition/Change
+
+ - Add a protocol specifier for rules to explicitly declare the desired protocol
+   for Ingress
+ - Add TCP protocol type to allow definition of TCP Ingress
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - protocol: tcp
+      port: 3306
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+
+    - protocol: http
+      host: www.foo.com
+      paths:
+      - path: /bar
+        backend:
+            serviceName: bar
+            servicePort: 80
+
+    - protocol: https
+      host: www.foo.com
+      tls:
+        - secretName: testsecret
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+```
+
+Although the proposal is to use a protocol specifier for http/https traffic
+it does not preclude supporting the http and tls ingress types from the
+existing Ingress spec.
+
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+    name: test-ingress
+spec:
+    rules:
+    - protocol: tcp
+      port: 3306
+      backend:
+        serviceName: testmysql
+        servicePort: 3306
+
+    - host: www.foo.com
+      http:
+        paths:
+        - path: /bar
+          backend:
+              serviceName: bar
+              servicePort: 80
+
+    - tls:
+        - secretName: testsecret
+      backend:
+        serviceName: bar
+        servicePort: 80
+```
+
+Despite this the proposal is to eventually deprecate the existance of a http or
+tls key as the indication of which protocol is desired for Ingress, and use an
+explicit protocol specifier.
+
+## Aditional notes about Cloud Providers' supported protocols
+
+### AWS
+
+AWS supports TCP
+
+#### Protocols
+
+ - HTTP
+ - HTTPS
+ - TCP
+ - TLS
+
+#### Security Groups
+
+Required for each LB instance
+
+#### Static IP
+
+No support for static IP
+
+### GCE
+
+#### Protocols
+
+ - HTTP
+ - HTTPS
+ - TCP
+ - UDP
+
+### Mesos
+
+### Openstack
+
+ - TCP
+ - UDP
+
+### Rackspace
+
+ - TCP
+ - UDP
+
+The following protocols are advertised but they seem like convenience options
+that map to a TCP or UDP port as appropriate, rather than actual support for
+the underlying protocol.
+
+ - DNS
+ - FTP
+ - HTTP
+ - HTTPS
+ - IMPAS
+ - IMPAv2
+ - IMPAv3
+ - IMPAv4
+ - LDAP
+ - LDAPS
+ - MYSQL
+ - POP3
+ - POP3S
+ - SFTP
+ - SMTP
+
```

Comment:
There's also `SMTPS`, [port 465 but it seems to have been deprecated](https://en.wikipedia.org/wiki/Simple_Mail_Transfer_Protocol#Ports).